### PR TITLE
feat: Integration for Chalice

### PR DESCRIPTION
**Pull request by Gleekzone**
This is a proposal for the integration of Chalice in Sentry

### CODE REVIEW

File: sentry_sdk/integrations/chalice.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+class ChaliceIntegration(Integration):
+    identifier = "chalice"
+
+    @staticmethod
+    def setup_once():
+        # for @app.route()
+        Chalice._get_view_function_response = _get_view_function_response
+        # for everything else (like events)
+        AwsLambdaIntegration.setup_once()
```
Comment (untitaker) [inline sentry_sdk/integrations/chalice.py:86]: Would just document that one needs to enable this. Unfortunately we have no way to enforce this programmatically right now. What you're doing here will enable the integration twice, meaning double the instrumentation overhead.

File: sentry_sdk/integrations/chalice.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        # type: (Event, Hint, datetime) -> Optional[Event]
+        extra = event.setdefault("extra", {})
+        extra["lambda"] = {
+            "function_name": lambda_context.function_name,
+            "function_version": lambda_context.function_version,
+            "invoked_function_arn": lambda_context.invoked_function_arn,
+            "aws_request_id": lambda_context.aws_request_id,
+        }
+
+        extra["cloudwatch logs"] = {
```
Comment (untitaker) [inline sentry_sdk/integrations/chalice.py:103]: This should not be necessary if the aws lambda integration is enabled

File: test-requirements.txt [hunk 3]
```diff
@@ -13,3 +13,5 @@ dnspython<2.0
 newrelic
 executing
 asttokens
+chalice
```
Comment (untitaker) [inline test-requirements.txt:16]: Please put this stuff into tox.ini, i.e. create a new job for your integration (+ testing different versions you want to support)

File: sentry_sdk/integrations/chalice.py (hunk 1 above)
Comment (Gleekzone) [reply sentry_sdk/integrations/chalice.py]: i actually found a better way to catch exceptions in events running in chalice, which doesn't require adding the aws lambda integration https://github.com/getsentry/sentry-python/blob/f1443129021cd1d4a61425b293eda49ec712200f/sentry_sdk/integrations/chalice.py#L113

File: sentry_sdk/integrations/chalice.py (hunk 2 above)
Comment (Gleekzone) [reply sentry_sdk/integrations/chalice.py]: this is now addressed by the change mentioned in the comment above https://github.com/getsentry/sentry-python/pull/779#discussion_r470299614

File: test-requirements.txt (hunk 3 above)
Comment (Gleekzone) [reply test-requirements.txt]: done!

File: sentry_sdk/integrations/chalice.py (hunk 2 above)
Comment (untitaker) [reply sentry_sdk/integrations/chalice.py]: Can we still deduplicate the logic of adding cloudwatch logs in both chalice and aws lambda? I.e. make sure that we write the same key in extra (so we don't accidentally send the same info twice)

Ideally you'd call the event processor from aws lambda in here.

File: sentry_sdk/integrations/chalice.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+            exc_info = sys.exc_info()
+            event, hint = event_from_exception(
+                exc_info,
+                client_options=client.options,
+                mechanism={"type": "chalice", "handled": False},
+            )
+            hub.capture_event(event, hint=hint)
+            hub.flush()
+            headers = {}
+            app.log.error("Caught exception for %s", view_function, exc_info=True)
```
Comment (untitaker) [inline sentry_sdk/integrations/chalice.py:81]: I really don't think the SDK is supposed to do that. Just re-raise the exception.

File: tox.ini [hunk 5]
```diff
@@ -69,8 +69,9 @@ envlist =
     {py2.7,py3.7,py3.8}-sqlalchemy-{1.2,1.3}
 
     py3.7-spark
-
+
```
Comment (untitaker) [inline tox.ini:72]: Please undo this change

File: sentry_sdk/integrations/__init__.py [hunk 6]
```diff
@@ -62,7 +62,7 @@ def iter_default_integrations(with_auto_enabling_integrations):
     "sentry_sdk.integrations.rq.RqIntegration",
     "sentry_sdk.integrations.aiohttp.AioHttpIntegration",
     "sentry_sdk.integrations.tornado.TornadoIntegration",
-    "sentry_sdk.integrations.sqlalchemy.SqlalchemyIntegration",
```
Comment (untitaker) [inline sentry_sdk/integrations/__init__.py:65]: Please undo this change

File: tests/integrations/chalice/test_chalice.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+from pytest_chalice.handlers import RequestHandler
+
+
+SENTRY_DSN = "https://111@sentry.io/111"
+
+
+@pytest.fixture
+def app():
+    sentry_sdk.init(dsn=SENTRY_DSN, integrations=[chalice_sentry.ChaliceIntegration()])
```
Comment (untitaker) [inline tests/integrations/chalice/test_chalice.py:18]: Use the `sentry_init` pytest fixture instead, then you dont need to pass a bogus DSN

File: sentry_sdk/integrations/chalice.py (hunk 4 above)
Comment (Gleekzone) [reply sentry_sdk/integrations/chalice.py:81]: i made a monkeypatch of the '_get_view_function_response' function from the chalice, in order to keep the uniformity of the code (a good practice of chalice), I keep this line https://github.com/aws/chalice/blob/2e44fe3b63378b488af88bd92254c58708cb9ac2/chalice/app.py#L1123

File: tox.ini (hunk 5 above)
Comment (Gleekzone) [reply tox.ini:72]: done!

File: sentry_sdk/integrations/__init__.py (hunk 6 above)
Comment (Gleekzone) [reply sentry_sdk/integrations/__init__.py:65]: done!

File: tests/integrations/chalice/test_chalice.py (hunk 7 above)
Comment (Gleekzone) [reply tests/integrations/chalice/test_chalice.py:18]: done!

File: sentry_sdk/integrations/chalice.py (hunk 2 above)
Comment (Gleekzone) [reply sentry_sdk/integrations/chalice.py]: This is resolved adding the event processor from aws lambda as you suggested me

File: sentry_sdk/integrations/chalice.py (hunk 4 above)
Comment (untitaker) [reply sentry_sdk/integrations/chalice.py:80]: Oh that's not great, copypasting the code. What about wrapping view_function, and passing the wrapped function into the original implementation of `_get_view_function_response`?

**Review by untitaker — CHANGES_REQUESTED** [review]
Can you resolve the last issue about copypasted code from chalice, and add a `__init__` in `tests/integrations/chalice` that calls `importorskip` (like in other integrations), otherwise the base testsuite does not pass (see travis CI)

File: sentry_sdk/integrations/chalice.py (hunk 4 above)
Comment (untitaker) [reply sentry_sdk/integrations/chalice.py:80]: Why don't you have access to the app context?

```python
def _get_view_function_response(app, view_function, ...):
    def new_view_function(**kw):
        # you can access app.lambda_context here
        try:
            return view_function(**kw)
        except Exception:
            ...

    return old_get_view_function_response(app, new_view_function, ...)

File: sentry_sdk/integrations/chalice.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                scope.transaction = app.lambda_context.function_name
+                scope.add_event_processor(
+                    _make_request_event_processor(
+                        app.current_request.to_dict(),
+                        app.lambda_context,
+                        configured_time,
+                    )
+                )
+            exc_info = sys.exc_info()
+            event, hint = event_from_exception(
```
Comment (untitaker) [inline sentry_sdk/integrations/chalice.py:72]: Please add the event processor unconditionally, not only when an exception is fired. If the user calls `capture_message` we also want data to be attached

Comment (Gleekzone) [reply sentry_sdk/integrations/chalice.py]: Done!

File: sentry_sdk/integrations/chalice.py (hunk 4 above)
Comment (Gleekzone) [reply sentry_sdk/integrations/chalice.py]: I solved the last problem about the code copied from chalice and this is my suggestion https://github.com/getsentry/sentry-python/blob/59f092d760da9ef888616e3c9cb7a53ece382f73/sentry_sdk/integrations/chalice.py#L48 and https://github.com/getsentry/sentry-python/blob/59f092d760da9ef888616e3c9cb7a53ece382f73/sentry_sdk/integrations/chalice.py#L69, could be another way to wrap the original function _get_view_function_response.

File: sentry_sdk/integrations/chalice.py [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+                    exc_info,
+                    client_options=client.options,
+                    mechanism={"type": "chalice", "handled": False},
+                )
+                hub.capture_event(event, hint=hint)
+                hub.flush()
+                reraise(*exc_info)
+
+
+old_get_view_function_response = Chalice._get_view_function_response
```
Comment (untitaker) [inline sentry_sdk/integrations/chalice.py:44]: Please don't have a global backup of the view function response, this will break if multiple libraries monkeypatch chalice

Comment (Gleekzone) [reply sentry_sdk/integrations/chalice.py]: add this in setup_once().

Comment (Gleekzone) [reply sentry_sdk/integrations/chalice.py]: Regarding the tests, these fail because the parameters of the context '_max_runtime' arrive null, so this line is not executed, the response if it contains the status_code '500' and the json, I don't understand why it fails in travis.

**Comment by untitaker** [comment]
Tried to fix conflicts but was denied pushing. Continuing at https://github.com/getsentry/sentry-docs/pull/2220

**Review by untitaker — APPROVED** [review]
almost there, please fix the linting errors (run `make lint` locally)

**Comment by matin** [comment]
@untitaker you now have permissions to push to our fork.

We're already using [`sentry-chalice`](https://github.com/cuenca-mx/sentry-chalice) in multiple projects. It'll be great switch to the official `sentry_sdk` library!

**Comment by matin** [comment]
fwiw, Chalice _does_ have a stub file for type hinting, [`app.pyi`](https://github.com/aws/chalice/blob/master/chalice/app.pyi), and includes [`py.typed`](https://github.com/aws/chalice/blob/master/chalice/py.typed).

It shouldn't be necessary to ignore the import to resolve this error: https://travis-ci.com/github/getsentry/sentry-python/jobs/382832642#L249-L251

**Comment by untitaker** [comment]
Yes but we can't install all optional dependencies at once when type
checking. Also we should not rely on those stubs as they are wrong (eg
flask, celery) or change between versions.

Technically we would have to typecheck for each entry in the cross product
of all versions of frameworks we support.
